### ISSUE TITLE
crash in twisted.web.http triggered by a blackberry web browser

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3601]
|<img alt="chbug's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| chbug reported|
|-|-|
|Trac ID|trac#3601|
|Type|defect|
|Created|2009-01-07 16:44:40Z|
|Branch|https://github.com/twisted/twisted/tree/invalid-if-modified-since-3601|

I've seen this in the logs of an app I wrote, which used to run twisted 2.4.0, but the problem seems to be present is 8.2.0 as well.

```
2009/01/07 14:51 GMT [HTTPChannel,709,127.0.0.1] 127.0.0.1 - - [07/Jan/2009:14:51:28 +0000] "GET /images/star-off.png HTTP/1.1" 500 1965 "http://xxx/xxx" "BlackBerry8820/4.2.2 Profile/MIDP-2.0 Configuration/CLDC-1.1 VendorID/118"
2009/01/07 14:51 GMT [HTTPChannel,709,127.0.0.1] Traceback (most recent call last):
          File "/usr/lib/python2.4/site-packages/twisted/internet/defer.py", line 307, in _runCallbacks
            self.result = callback(self.result, *args, **kw)
          File "/usr/lib/python2.4/site-packages/nevow/appserver.py", line 157, in gotPageContext
            return defer.maybeDeferred(
          File "/usr/lib/python2.4/site-packages/twisted/internet/defer.py", line 107, in maybeDeferred
            result = f(*args, **kw)
          File "/usr/lib/python2.4/site-packages/nevow/appserver.py", line 361, in renderHTTP
            result = defer.maybeDeferred(self.original.render, request).addCallback(
        --- <exception caught here> ---
          File "/usr/lib/python2.4/site-packages/twisted/internet/defer.py", line 107, in maybeDeferred
            result = f(*args, **kw)
          File "/usr/lib/python2.4/site-packages/twisted/web/static.py", line 325, in render
            if request.setLastModified(self.getmtime()) is http.CACHED:
          File "/usr/lib/python2.4/site-packages/twisted/web/http.py", line 799, in setLastModified
            modified_since = stringToDatetime(modified_since.split(';', 1)[0])
          File "/usr/lib/python2.4/site-packages/twisted/web/http.py", line 296, in stringToDatetime
            month = int(monthname_lower.index(month.lower()))
        exceptions.ValueError: list.index(x): x not in list
```

I suspect the browser to send some funny date in the request, but I have no way to reproduce it. The code definitely doesn't expect weird month names. I'll try to provide a patch if you don't beat me to it :)

<details><summary>Searchable metadata</summary>

```
trac-id__3601 3601
type__defect defect
reporter__chbug chbug
priority__normal normal
milestone__ 
branch__branches_invalid_if_modified_since_3601 branches/invalid-if-modified-since-3601
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1231346680000000 1231346680000000
changetime__1231532329000000 1231532329000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun set owner to @exarkun|
|-|-|

Some quick googling reveals various complaints in forums that Blackberry requests use localized day-of-week and month names in the dates they generate.  This is, of course, broken.  We can handle the error and ignore the value of the If-Modified-Since header easily enough, though.

For reference, this appears to be an example of a German Blackberry's If-Modified-Since header value: _Fr, 31 Okt 2008 21:26:30 GMT_

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

(In [25865]) Branching to 'invalid-if-modified-since-3601'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

(In [25866]) handle invalid if-modified-since requests by ignoring the if-modified-since value

refs [#3601](https://github.com/twisted/twisted/issues/3601)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

please review

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve set owner to @exarkun|
|-|-|

Just one comment:
```
+            try:
+                modifiedSince = stringToDatetime(firstPart)
+            except:
+                return None
```

Can we be more strict in the except here? I guess `ValueError` and `TypeError` should be enough, no?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun removed owner|
|-|-|

You're right.  I was lazy and didn't want to figure out what exceptions `stringToDatetime` might raise.  It seems like `ValueError` and `AssertionError`, so I added tests for those cases and changed the assert to raise a `ValueError` instead.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve set owner to @exarkun|
|-|-|

Nice, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set status to `closed`|
|-|-|

(In [25934]) Merge invalid-if-modified-since-3601 

Author: exarkun
Reviewer: therve
Fixes: [#3601](https://github.com/twisted/twisted/issues/3601)

Handle exceptions raised while parsing the value of the If-Modified-Since
header by treating the header as though it were not present.  This lets
the server generate responses to requests which include an invalid value
for this header.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
